sdk/typescript/runtime/main.go

```diff
@@ -50,17 +57,34 @@ func (t *TypeScriptSdk) ModuleRuntime(ctx context.Context, modSource *ModuleSour
 		return nil, fmt.Errorf("could not load module config: %v", err)
 	}
 
+	detectedRuntime, err := t.DetectRuntime(ctx, modSource, subPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create module runtime: %w", err)
+	}
+
 	entrypointPath := path.Join(ModSourceDirPath, subPath, EntrypointExecutablePath)
 	tsConfigPath := path.Join(ModSourceDirPath, subPath, "tsconfig.json")
-	return ctr.
+
+	ctr = ctr.WithMountedFile(entrypointPath, ctr.Directory("/opt/bin").File(EntrypointExecutableFile))
+
+	switch detectedRuntime {
+	case Bun:
+		return ctr.
+			// Install dependencies
+			WithExec([]string{"bun", "install"}).
+			WithEntrypoint([]string{"bun", entrypointPath}), nil
+	case Node:
+		return ctr.
 		// Install dependencies
 		WithExec([]string{"npm", "install"}).
-		WithMountedFile(entrypointPath, ctr.Directory("/opt/bin").File(EntrypointExecutableFile)).
 		// need to specify --tsconfig because final runtime container will change working directory to a separate scratch
 		// dir, without this the paths mapped in the tsconfig.json will not be used and js module loading will fail
 		// need to specify --no-deprecation because the default package.json has no main field which triggers a warning
 		// not useful to display to the user.
 		WithEntrypoint([]string{"tsx", "--no-deprecation", "--tsconfig", tsConfigPath, entrypointPath}), nil
+	default:
+		return nil, fmt.Errorf("unknown runtime: %v", detectedRuntime)
+	}
```

Comment:
If bun is faster I'm ok making that the default instead of Node (dagger init), just not sure how the community will receive that, it takes one to know one.

I'm considering also in Python adopting a new tool as a default since it's so much faster. We need to make it as fast as possible to get to the function to get a more responsive CLI and avoid adding bottlenecks in the ecosystem.

Reply:
I'm also not opposed to switching the default, but I think we should aim to land that change in a separate release from the one we first introduce it in - introducing and *potentially* breaking in the same release is something I'd like to avoid.

Reply:
While I also think switching the default, just for the speed is generally a good idea; we cannot forget that most people assume node is the default runtime and while bun works with the dagger SDK not all node APIs and therefore not all npm packages work with bun yet. Most popular packages do work, but if someone uses a non-compatible npm package, the resulting error might be very confusing. (Example: jsonwebtoken uses nodes Diffie-Hellman, which is not implemented yet in bun)
This might be something we can alleviate with good documentation but still worth considering.

Reply:
I don't want to sound like the pesimist but I'd strongly advocate to **not** changing the default runtime to bun.  As @benjaminstrasser mentioned above, bun is still not fully node API compatible and there are some libraries that fail wiredly when using it

Reply:
> `DetectedRuntime` already validates supported runtimes so this will never run. If defaulting to Node, it would be more like:
> 
> ```go
> if isBun {
>   // do bun
> } else {
>   // do node
> }
> ```
> 
> And later:
> 
> ```go
> switch runtime {
> case Bun:
>     ...
> case Deno:
>     ...
> default:
>     // node
> }
> ```
>

I am usually trying to be rather defensive with implicit guarantees. If the compiler cannot assure something, which is the case as DetectRuntime can return `SupportedTSRuntime("bla)"`, I am checking it at runtime. I usually prefer that the consumer of  function can only trust explicit compile-time guarantees . F.e in TS or Java allow for such compile time guarantees, but I have not found a way to do this in go. Maybe I am missing something here.

I am also completely fine with changing it however. Just a personal preference of mine.

> Another way to organize this is to implement an interface (just in Go, not exposed to API) and create different `type Node struct` and `type Bun struct` that implement that interface, returned by `DetectRuntime() Runtime` (standalone function, not a part of `TypeSriptSdk` as to not expose). This is what I plan on doing with Python.

Would you like me to look into it now or replicate your design approach for ts once you have a draft for python?

Reply:
> I am usually trying to be rather defensive with implicit guarantees. If the compiler cannot assure something, which is the case as DetectRuntime can return `SupportedTSRuntime("bla)"`, I am checking it at runtime. I usually prefer that the consumer of function can only trust explicit compile-time guarantees . F.e in TS or Java allow for such compile time guarantees, but I have not found a way to do this in go. Maybe I am missing something here.

It's more like, node is what we support up to now. But... if you have `"runtime": "bun"`, you can use Bun instead. If it's `"runtime": "bla"` then sure, return an error. But If it's empty, and you don't auto detect from a bun lock file, rather than looking for a node lock file, just assume node.

You can argue that adding the check for node as well can give a more useful error earlier to the user in case things get messed up somehow. So I'm not opposed to it.

> Would you like me to look into it now or replicate your design approach for ts once you have a draft for python?

A refactor isn't necessary for the PR's approval (more important is the performance concerns), unless you want to or you're inspired by the following ideas that I'm working on:
- A `Config` struct to hold configuration, instantiated in the constructor with the SDK's defaults, corresponding to the data structure in `package.json`'s `"dagger"` key;
- Using `modSource.Directory("").Entries(ctx)` to do some discoverability over which files are present (like the bun lock file) to override the defaults in the initial config, but also for checking if there's a `package.json` to begin with which replaces the WithExec that copies from the template if it doesn't exist;
- Unmarshalling `"dagger"` field from `package.json` into the `Config` instance, for explicit user settings;
- Using the final config to build the runtime Container ;